core/block.js

```diff
@@ -668,6 +672,57 @@ Blockly.Block.prototype.setMovable = function(movable) {
   this.movable_ = movable;
 };
 
+/**
+ * Get whether is block is duplicatable or not. If any decendent is not
+ * duplicatable this block is not duplicatable. If duplicating this block and
+ * decendents will put this block over the workspace\'s capacity this block is
+ * not duplicatable. If duplicating this block and decendents will put any
+ * decendent over their maxInstances this block is not duplicatable.
+ * @return {boolean} True if duplicatable
+ */
+Blockly.Block.prototype.isDuplicatable = function() {
+  var instancesOfType = this.workspace.getBlocksByType(this.type);
+  if (instancesOfType.length >= this.maxInstances) {
+    return false;
+  }
+
+  var copyableBlocks = this.getDescendants(true);
+  // Remove all "next statement" blocks because they will not be copied.
+  if (this.getNextBlock()) {
+    var index = copyableBlocks.indexOf(this.getNextBlock());
+    copyableBlocks.splice(index, copyableBlocks.length - index);
+  }
+  
+  if (copyableBlocks.length > this.workspace.remainingCapacity()) {
+    return false;
+  }
+
+  // If we do not have any blocks to copy besides ourself, we can just
+  // return true, because we already checked ourself.
+  if (copyableBlocks.length > 1) {
+    var type = this.type;
+    var copyableBlocksOfType = copyableBlocks.filter(function(block) {
```

Comment:
I'm not sure this will correctly count all types in the entire stack. If there are multiple branches (if/else block) with limited blocks in each branch I think you'll only get one branch worth at a time.

Test case:
Foo block has maxInstances of 3.
Workspace contains an If/Else block.
 - If branch contains block Foo.
 - Else branch contains block Foo.

Reply:
You were correct! They logic I had in place didn't cover sibling relationships. It now runs through all the copyable blocks & checks if there is enough "remainingCapacity" for each type.